presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LeftJoinWithArrayContainsToEquiJoinCondition.java

```diff
@@ -0,0 +1,173 @@
+ *     - Left Join
+ *          l_key = field
+ *          - scan l
+ *          - Unnest
+ *              field <- unnest distinct_array
+ *              - project
+ *                  distinct_array := filter(array_distinct(r_array), element -> element is not null)
+ *                  - scan r
+ *                      r_array
+ * </pre>
+ */
+public class LeftJoinWithArrayContainsToEquiJoinCondition
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join().matching(x -> x.getType().equals(JoinNode.Type.LEFT) && x.getCriteria().isEmpty() && x.getFilter().isPresent());
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final RowExpressionDeterminismEvaluator determinismEvaluator;
+    private final FunctionResolution functionResolution;
+
+    public LeftJoinWithArrayContainsToEquiJoinCondition(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.determinismEvaluator = new RowExpressionDeterminismEvaluator(functionAndTypeManager);
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRewriteLeftJoinArrayContainsToEquiJoinEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        RowExpression filterPredicate = node.getFilter().get();
+        List<VariableReferenceExpression> leftInput = node.getLeft().getOutputVariables();
+        List<VariableReferenceExpression> rightInput = node.getRight().getOutputVariables();
+        List<RowExpression> andConjuncts = extractConjuncts(filterPredicate);
+        List<RowExpression> remainingConjuncts = ImmutableList.of();
+        RowExpression array = null;
+        RowExpression element = null;
+
+        for (RowExpression conjunct : andConjuncts) {
+            if (conjunct instanceof CallExpression && functionResolution.isArrayContainsFunction(((CallExpression) conjunct).getFunctionHandle())) {
+                RowExpression arrayExpression = ((CallExpression) conjunct).getArguments().get(0);
+                RowExpression elementExpression = ((CallExpression) conjunct).getArguments().get(1);
+                if (determinismEvaluator.isDeterministic(arrayExpression) && rightInput.containsAll(extractAll(arrayExpression))
+                        && determinismEvaluator.isDeterministic(elementExpression) && leftInput.containsAll(extractAll(elementExpression))) {
+                    array = arrayExpression;
+                    element = elementExpression;
+                    remainingConjuncts = andConjuncts.stream().filter(x -> !x.equals(conjunct)).collect(toImmutableList());
```

Comment:
this looks like we're doing a n^2 computation: scanning andConjuncts again in a for-loop on top of andConjuncts.
I know that the break below makes it linear and non-quadratic but I think it is confusing and less readable that way.

Reply:
Done